### ISSUE TITLE
[FEAT] Analyse blocking rule skew by finding the worst n offenders

### ISSUE THREAD

**Issue by RobinL** [moj-analytical-services/splink#1946]
At the moment you can have a blocking rule that works well for almost all your data, but it affected by data skew.

For example, a first name and surname blocking rule may be affected by the 'John Smith' problem, and in turn this can cause Spark jobs to fail

We could have a function that detected these 'worst offenders'

Very rough implementation below:

<details>
<summary>
script to achieve this
</summary>

```python
import duckdb

from splink.blocking import (
    BlockingRule,
)
from splink.datasets import splink_datasets
from splink.duckdb.blocking_rule_library import block_on
from splink.duckdb.linker import DuckDBLinker

df = splink_datasets.historical_50k
df.head()

settings = {
    "link_type": "link_only",
    "blocking_rules_to_generate_predictions": [
        block_on(["first_name", "dob"]),
        block_on(["first_name", "surname"]),
        block_on(["surname", "dob", "birth_place"]),
    ],
}


linker = DuckDBLinker(df, settings)


def count_worst_block_skew(linker, blocking_rule):
    if isinstance(blocking_rule, str):
        blocking_rule = BlockingRule(blocking_rule, sqlglot_dialect=linker._sql_dialect)

    join_conditions = blocking_rule._equi_join_conditions

    l_cols_sel = []
    r_cols_sel = []
    l_cols_gb = []
    r_cols_gb = []
    using = []
    for (
        i,
        (l_key, r_key),
    ) in enumerate(join_conditions):
        l_cols_sel.append(f"{l_key} as key_{i}")
        r_cols_sel.append(f"{r_key} as key_{i}")
        l_cols_gb.append(l_key)
        r_cols_gb.append(r_key)
        using.append(f"key_{i}")

    l_cols_sel_str = ", ".join(l_cols_sel)
    r_cols_sel_str = ", ".join(r_cols_sel)
    l_cols_gb_str = ", ".join(l_cols_gb)
    r_cols_gb_str = ", ".join(r_cols_gb)
    using_str = ", ".join(using)

    sqls = []

    if linker._two_dataset_link_only:
        #    Can just use the raw input datasets
        keys = list(linker._input_tables_dict.keys())
        input_tablename_l = linker._input_tables_dict[keys[0]].physical_name
        input_tablename_r = linker._input_tables_dict[keys[1]].physical_name

    else:
        input_tablename_l = "__splink__df_concat"
        input_tablename_r = "__splink__df_concat"

    if not join_conditions:
        if linker._two_dataset_link_only:
            sql = f"""
            SELECT
                (SELECT COUNT(*) FROM {input_tablename_l})
                *
                (SELECT COUNT(*) FROM {input_tablename_r})
                    AS count_of_pairwise_comparisons_generated
            """
        else:
            sql = """
            select count(*) * count(*) as count_of_pairwise_comparisons_generated
            from __splink__df_concat

            """
        sqls.append(
            {"sql": sql, "output_table_name": "__splink__total_of_block_counts"}
        )
        return sqls

    sql = f"""
    select {l_cols_sel_str}, count(*) as count_l
    from {input_tablename_l}
    group by {l_cols_gb_str}
    """

    sqls.append(
        {"sql": sql, "output_table_name": "__splink__count_comparisons_from_blocking_l"}
    )

    sql = f"""
    select {r_cols_sel_str}, count(*) as count_r
    from {input_tablename_r}
    group by {r_cols_gb_str}
    """

    sqls.append(
        {"sql": sql, "output_table_name": "__splink__count_comparisons_from_blocking_r"}
    )

    sql = f"""
    select *, count_l, count_r, count_l * count_r as block_count
    from __splink__count_comparisons_from_blocking_l
    inner join __splink__count_comparisons_from_blocking_r
    using ({using_str})
    """

    sqls.append({"sql": sql, "output_table_name": "__splink__block_counts"})

    sql = """
    select *
    from __splink__block_counts
    order by block_count desc
    limit 3
    """

    sqls.append({"sql": sql, "output_table_name": "__splink__total_of_block_counts"})

    return sqls


for br in settings["blocking_rules_to_generate_predictions"]:

    sqls = count_worst_block_skew(linker, br)

    input_dataframes = []
    df_concat = linker._initialise_df_concat()

    if df_concat:
        input_dataframes.append(df_concat)
    for sql in sqls:
        linker._enqueue_sql(sql["sql"], sql["output_table_name"])
    print("----")
    print(br)

    display(linker._execute_sql_pipeline(input_dataframes).as_pandas_dataframe())

```

</details>

![image](https://github.com/moj-analytical-services/splink/assets/2608005/ea584662-7e93-489c-8110-c015751a2423)

**Comment by aleks-drozy** [comment]
Looked into this before starting work, and I think the core ask here is already done: `n_largest_blocks` (added in #2260, `splink/internals/blocking_analysis.py`) does exactly what's described - given a blocking rule, finds the key-value combinations producing the largest blocks, e.g. the "John Smith" case in its own docstring. It's public (`splink.n_largest_blocks`), tested, and documented. Since this issue has no linked PR, I think it just never got closed out.

I hit the exact failure mode this issue describes at real scale in my own project (entity resolution across ~900k company records): a `legal_form` blocking rule that was ~92% concentrated on one value ("LTD"), producing an estimated 23 billion comparisons and hanging for 3+ hours before I caught it. So while building a small repro to confirm `n_largest_blocks` would have caught that, I found two real gaps worth a follow-up:

1. **Bug:** `n_largest_blocks` crashes with a raw, cryptic SQL parser error (`syntax error at or near ","`) on any blocking rule with no equi-join conditions (pure filter rules like `levenshtein(...) <= 1`, or `Or(...)` combinations) - the empty key list produces invalid SQL (`select , count_l ... using ()`). Should raise a clear `ValueError` explaining why instead. I'll open a small PR for this.

2. **Enhancement:** the output has raw `block_count` per block, but no share of the rule's total comparisons - which was the actual number that mattered in my incident ("this one value is 92% of everything"). Adding `proportion_of_comparisons` (+ a cumulative column) would make that read off directly instead of requiring a second `count_comparisons_from_blocking_rules` call and manual division.

Happy to open a draft PR for #2 too, but since it changes a public function's output schema, wanted to ask first: would you rather that be unconditional (new columns always present) or opt-in via a kwarg? Can go either way.
